sklearn/cluster/_optics.py

```diff
@@ -665,10 +665,14 @@ def _set_reach_dist(
 
     # Only compute distances to unprocessed neighbors:
     if metric == "precomputed":
-        dists = X[point_index, unproc]
-        if issparse(dists):
-            dists.sort_indices()
-            dists = dists.data
+        dists = X[[point_index], unproc]
+        if isinstance(dists, np.matrix):
+            # X is a scipy sparse matrix.
+            dists = np.asarray(dists)
+        elif hasattr(dists, "toarray"):
+            # X is a scipy sparse array.
+            dists = dists.toarray()
```

Comment:
@marenwestermann do you understand why this branch of the condition is not already covered by `test_precomputed_dists`?

I am a bit confused.

Reply:
@ogrisel The issue is that previously `dists` was a sparse matrix for the test case `"is_sparse"=True`. Now, when we want to test for the case `csr_matrix`, `dists` becomes a numpy matrix (which is currently covered in the code) and in the case where we want to test for `csr_array`, `dists` becomes a numpy array. So we never test for the case sparse array. 

Now the question is should `dists` be sparse when X is sparse (as it was preciously the case) or is the current implementation ok? I don't have a deep enough understanding of the OPTICS estimator to make a judgement.

Reply:
Just a minimal example to understand what happens here:

```python
In [5]: from scipy import sparse

In [6]: X = sparse.random(10, 10, format="csr")

In [7]: X[0, :]
Out[7]: 
<1x10 sparse matrix of type '<class 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>

In [8]: X[0, [1, 2]]
Out[8]: 
<1x2 sparse matrix of type '<class 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>

In [9]: X[[0], [1, 2]]
Out[9]: matrix([[0., 0.]])
```

Since we index both dimension, we get this `np.matrix` that is the difference with the previous code.

For me, we don't need to cover anymore the case where we would eventually have a sparse matrix because we will get already a `np.matrix` or a `np.array`.

Reply:
@marenwestermann do you think that you can reduce to only having `np.matrix` or `np.array` (not anymore sparse matrices) for the condition. Once done, I think that we good for merging.